Fix Issue 17998 - Document Options for install.sh

wilzbach opened the pull request:
While this isn't perfect, it should provide a good basis for continuous improvements ;-)

I wasn't sure whether `~/dlang/install.sh` or `./install.sh` should be the recommended invocation.

CC @MartinNowak @ZombineDev

ZombineDev commented:
Thanks, this was on my to do list as well ;)

ZombineDev commented on install.dd:
```diff
@@ -0,0 +1,163 @@
+$(H2 $(LNAME2 usage, Usage))
+
+$(CONSOLE
+./install.sh [&lt;command&gt;] [&lt;args&gt;]
```
Always use `~/dlang/install.sh`. Avoid using relative paths, especially in documentation.

wilzbach commented:
> What do you think about an introductory paragraph, along the lines of:

I like it. I converted your text to Ddoc and made a few changes. Please feel free to improve upon them ;-)

andralex commented on install.dd:
```diff
@@ -0,0 +1,200 @@
+If you want, you can directly invoke it:
+
+$(CONSOLE
+curl https://dlang.org/install.sh | bash -s
```
I recall someone (@CyberShadow?) quoted a scathing article about this practice. Perhaps it would be best to not mention it.

CyberShadow replied:
AFAIK, it's fine if done right:
https://sandstorm.io/news/2015-09-24-is-curl-bash-insecure-pgp-verified-install

wilzbach replied:
Well we do mention it on the downloads page too:

![image](https://user-images.githubusercontent.com/4370550/34091664-79d16c34-e3bf-11e7-92e3-b9eed1c4c90b.png)

andralex commented on install.dd:
```diff
@@ -0,0 +1,200 @@
+wget https://dlang.org/install.sh
+)
+
+If you want, you can directly invoke it:
```
Oh, also please avoid the use of "you" in documentation.

s/If you want, you can directly invoke it/Alternatively, the script can be invoked directly/

andralex commented on install.dd:
```diff
@@ -0,0 +1,200 @@
+
+$(P Once a compiler can be activated for the current session:)
+
+Run the respective `activate` script in your shell to use a compiler by default:
```
s/your shell/the shell/

wilzbach commented:
Okay addressed the feedback.
I also switched to use the [recently introduced](https://github.com/dlang/dlang.org/pull/1697) "info on hover" button on the downloads page to link to the script documentation (instead of simply using "Explain").

ZombineDev commented on install.dd:
```diff
@@ -0,0 +1,200 @@
+$(EXAMPLES
+~/dlang/install.sh uninstall dmd
+/install.sh uninstall dmd-2.071.1
+/install.sh uninstall ldc-1.1.0-beta2
```
Use `~/dlang/install.sh` here too. It looks a bit confusing, otherwise.

ZombineDev commented on install.dd:
```diff
@@ -0,0 +1,200 @@
+Run the respective `activate` script in a shell to use a compiler by default:
+
+$(CONSOLE
+source ~/dlang/&lt;installed-compiler&gt;/activate
```
You should also mention that for `fish` shell the command is:
```
source ~/dlang/&lt;installed-compiler&gt;/activate.fish
```

ZombineDev commented on install.dd:
```diff
@@ -0,0 +1,200 @@
+~/dlang/install.sh dmd
+~/dlang/install.sh install dmd
+~/dlang/install.sh install dmd-2.071.1
+~/dlang/install.sh install ldc-1.1.0-beta2
```
Just for completeness (and easy copy-pasting) can you mention also:
```
~/dlang/install.sh install dmd-beta
~/dlang/install.sh install dmd-nightly
```

wilzbach replied:
Ok. done

joakim-noah commented on install.dd:
```diff
@@ -0,0 +1,200 @@
+            Installs the latest beta version of a compiler)
+        $(SWITCH $(SWNAME dmd-nightly),
+            Installs DMD nightly)
+        $(SWITCH $(SWNAME dmd-20170-02-10),
```
2017-

joakim-noah commented on install.dd:
```diff
@@ -0,0 +1,200 @@
+~/dlang/install.sh install ldc-1.1.0-beta2
+)
+
+$(P Once a compiler can be activated for the current session:)
```
Why put this in its own paragraph?  Merge it with the next sentence instead `session, run the`

wilzbach commented on install.dd:
```diff
@@ -0,0 +1,200 @@
+ $(LI Maintaining one or more portable compiler installations in the user's home directory (`~/dlang`))
+ $(LI Maintaining the latest stable version of $(LINK2 http://code.dlang.org/docs/commandline, `dub`))
+ $(LI Self-updating via `~/dlang/install.sh update`)
+ $(LI Supporting nested activation of compilers. Activating a compiler, will setup the `PATH`, `LIBRARY_PATH`, `LD_LIBRARY_PATH`, `DMD`, `DC`, and `PS1` environment variables. After activation, user's can revert those variables to their old values by invoking the `deactivate` command, added to current shell session, by the activation command.)
```
> Can you activate a compiler inside another one's session?

Yes it's a stack.

wilzbach commented:
@ZombineDev @joakim-noah thanks a lot for your feedback! Addressed.

wilzbach commented on install.dd:
```diff
@@ -0,0 +1,204 @@
+$(H2 $(LNAME2 get, Downloading the installer))
+
+$(CONSOLE
+wget https://dlang.org/install.sh
```
How about:

```shell
mkdir -p ~/dlang && wget https://dlang.org/install.sh -O ~/dlang/install.sh
```

joakim-noah replied:
Yes, better to have a consistent install location, if the curl method installs it differently.

wilzbach commented on install.dd:
```diff
@@ -0,0 +1,210 @@
+~/dlang/&lt;installed-ldc-compiler&gt;/ldc2
+)
+
+Users of the $(LINK2 https://fishshell.com, Fish shell), can `activate.fish`:
```
I put the note about the FISH shell at the end of this section as AFAICT it's not a very popular shell and it stops the reading flow awkwardly if put in-between.

joakim-noah replied:
No comma after `shell),`

joakim-noah commented on install.dd:
```diff
@@ -0,0 +1,210 @@
+ $(LI Maintaining one or more portable compiler installations in the user's home directory (`~/dlang`))
+ $(LI Maintaining the latest stable version of the D package manager, $(LINK2 http://code.dlang.org/docs/commandline, `dub`))
+ $(LI Self-updating via `~/dlang/install.sh update`)
+ $(LI Supporting nested activation of compilers. Activating a compiler, will setup the `PATH`, `LIBRARY_PATH`, `LD_LIBRARY_PATH`, `DMD`, `DC`, and `PS1` environment variables, and a `deactivate` command, which reverts the modified variables to their old values.)
```
No comma needed after `compiler,` and `variables,`

wilzbach replied:
You are absolutely right about the first occurrence, but isn't the second one the Oxford comma?
I reworded it slightly:

> will setup the `PATH`, `LIBRARY_PATH`, .. `PS1` environment variables,  and a `deactivate` command

joakim-noah commented on install.dd:
```diff
@@ -0,0 +1,210 @@
+ $(LI Maintaining the latest stable version of the D package manager, $(LINK2 http://code.dlang.org/docs/commandline, `dub`))
+ $(LI Self-updating via `~/dlang/install.sh update`)
+ $(LI Supporting nested activation of compilers. Activating a compiler, will setup the `PATH`, `LIBRARY_PATH`, `LD_LIBRARY_PATH`, `DMD`, `DC`, and `PS1` environment variables, and a `deactivate` command, which reverts the modified variables to their old values.)
+ $(LI Supporting Fish shell, in addition to Bash-compatible shells.)
```
`the Fish`

joakim-noah commented on install.dd:
```diff
@@ -0,0 +1,210 @@
+)
+
+Alternatively, it is also possible to directly use a compiler by calling its
+binary directly.
```
Replace end with `possible to call the compiler binary directly`.

JinShil commented:
This is really good, but we are introducing a maintenance burden between install.sh and this file.  Would it not be better to add this documentation to the `--help` output of install.sh and maybe generate this file as part of the build?

wilzbach commented:
> but we are introducing a maintenance burden and a fragile synchronization between this file and install.sh. Would it not be better to add this documentation to the --help output of install.sh and maybe generate this file as part of the build?

Of course it would be better, but that's pretty difficult to do all the custom text, explanations and formatting.
I don't think that this file changes much (the installer has about 2-3 pull requests per month and they are mostly just upgrades to the release build system), so that it's worth looking into the effort of generating the page automatically.

OTOH the compiler help page changes quite a lot and we are incredibly bad in keeping it up to date. So if we want to invest time in generating something automatically, I think the `dmd-*.html` pages are a better candidate.

wilzbach commented:
> maybe linux shouldn't be capitalized

Hmm, I have only seen it written capitalized except for DMD's inconsistent version file.

(Also I am terribly sorry for forgetting to squash this. We really need to re-enable auto-merge-squash.)

MartinNowak commented:
> Would it not be better to add this documentation to the --help output of install.sh and maybe generate this file as part of the build?

Indeed figuring out how we can base those things on a common man-page like document would be helpful, but not that urgent as it's rarely changing.
https://github.com/fish-shell/fish-shell/blob/f563262cfda72aa291dcca4cd703b2ccba3331a9/share/tools/create_manpage_completions.py
What could be done fairly easily is dumping the outputs of `./install.sh --help, ./install.sh <cmd> --help` verbatim into a code block, but it's less versatile than the docs in *this* PR.
